File: python/007LogicOperator.py
'''
@Description: 像素的逻辑运算
@Author: shuiyj
@Date: 2019-07-23 17:21:28
@LastEditTime: 2019-07-23 18:59:13
'''
import cv2 as cv
import numpy as np

# create image src1
src1 = np.zeros(shape=[500, 500, 3], dtype=np.uint8)
src1[100:200, 100:200, 1] = 255
src1[100:200, 100:200, 2] = 255
cv.imshow("src1", src1)

# create image src2
src2 = np.zeros(shape=[500, 500, 3], dtype=np.uint8)
src2[150:250, 150:250, 1] = 255
cv.imshow("src2", src2)

# 逻辑运算
dst1 = cv.bitwise_and(src1, src2)
dst2 = cv.bitwise_or(src1, src2)
dst3 = cv.bitwise_xor(src1, src2)

cv.imshow("dst1", dst1)
cv.imshow("dst2", dst2)
cv.imshow("dst3", dst3)

# 取反操作
src = cv.imread("../images/1.png")
cv.namedWindow("src", cv.WINDOW_AUTOSIZE)
cv.imshow("src", src)
dst = cv.bitwise_not(src)

cv.imshow("dst", dst)

cv.waitKey(0)
cv.destroyAllWindows()

File: cpp/007LogicOperator.cpp
/*
 * @Description: 像素的逻辑运算
 * @Author: shuiyj
 * @Date: 2019-07-23 11:46:10
 * @LastEditTime: 2019-07-23 18:59:20
 */
#include <iostream>
#include <opencv2/opencv.hpp>

int main(int argc, char const *argv[])
{
    // create image src1
    cv::Mat src1 = cv::Mat::zeros(cv::Size(500, 500), CV_8UC3);
    cv::Rect rect(100, 100, 100, 100);
    src1(rect) = cv::Scalar(0, 0, 255);
    cv::imshow("src1", src1);

    // create image src2
    cv::Mat src2 = cv::Mat::zeros(cv::Size(500, 500), CV_8UC3);
    rect.x = 150;
    rect.y = 150;
    src2(rect) = cv::Scalar(0, 255, 255);
    cv::imshow("src2", src2);

    // 逻辑操作
    cv::Mat dst1, dst2, dst3;
    cv::bitwise_and(src1, src2, dst1);
    cv::bitwise_xor(src1, src2, dst2);
    cv::bitwise_or(src1, src2, dst3);

    cv::imshow("dst1", dst1);
    cv::imshow("dst2", dst2);
    cv::imshow("dst3", dst3);

    // 取反操作
    cv::Mat src = cv::imread("../images/1.png");
    cv::Mat dst;

    cv::bitwise_not(src, dst);

    cv::imshow("src", src);
    cv::imshow("dst", dst);

    cv::waitKey(0);

    return 0;
}
